src/openms/source/ANALYSIS/OPENSWATH/IonMobilityScoring.cpp

```diff
@@ -0,0 +1,476 @@
+
+   @param profile The ion mobility data
+   @param im_grid The grid to be used
+   @param al_int_values The intensity vector (y)
+   @param al_im_values The ion mobility vector (x)
+   @param max_peak_idx The grid position of the maximum
+
+  */
+  void alignToGrid(const IonMobilogram& profile,
+               const std::vector<double>& im_grid,
+               std::vector< double >& al_int_values,
+               std::vector< double >& al_im_values,
+               Size & max_peak_idx)
+  {
+    auto pr_it = profile.begin();
+    max_peak_idx = 0;
+    double max_int = 0;
+    for (Size k = 0; k < im_grid.size(); k++)
+    {
+      // In each iteration, the IM value of pr_it should be equal to or
+      // larger than the master container. If it is equal, we add the current
+      // data point, if it is larger we add zero and advance the counter k.
+      if (pr_it != profile.end() && fabs(pr_it->first - im_grid[k] ) < 1e-4 ) 
+      {
+        al_int_values.push_back(pr_it->second);
+        al_im_values.push_back(pr_it->first);
+        ++pr_it;
+      }
+      else
+      {
+        al_int_values.push_back(0.0);
+        al_im_values.push_back( im_grid[k] );
+      }
+      // OPENMS_LOG_DEBUG << "grid position " << im_grid[k] << " profile position " << pr_it->first << std::endl;
+
+      // check that we did not advance past 
+      if (pr_it != profile.end() && (im_grid[k] - pr_it->first) > 1e-3)
+      {
+        std::cout << " This should never happen, pr_it has advanced past the master container: " << im_grid[k]  << "  / " <<  pr_it->first  << std::endl;
+        throw Exception::OutOfRange(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION);
+      }
+
+      // collect maxima
+      if (pr_it != profile.end() && pr_it->second > max_int)
+      {
+        max_int = pr_it->second;
+        max_peak_idx = k;
+      }
+    }
+  }
+
+  /**
+    @brief Integrate intensity in an ion mobility spectrum from start to end
+
+    This function will integrate the intensity in a spectrum between mz_start
+    and mz_end, returning the total intensity and an intensity-weighted drift
+    time value.
+
+    This function also returns the full ion mobility profile in "res".
+
+    @note If there is no signal, mz will be set to -1 and intensity to 0
+  */
+  void integrateDriftSpectrum(OpenSwath::SpectrumPtr spectrum, 
+                              double mz_start,
+                              double mz_end,
+                              double & im,
+                              double & intensity,
+                              IonMobilogram& res, 
+                              double drift_start,
+                              double drift_end)
+  {
+    OPENMS_PRECONDITION(spectrum->getDriftTimeArray() != nullptr, "Cannot filter by drift time if no drift time is available.");
+
+    // rounding multiplier for the ion mobility value
+    // TODO: how to improve this -- will work up to 42949.67296
+    double IM_IDX_MULT = 10e5;
```

Comment:
maybe add more comment for projection to ints?
Are there problems with very similar values to spread over bins? i.e. for binned spectra some offsets can be used to reduce the effect. does something similar apply here?

Reply:
yes that is basically the idea, we assume that the ion mobility values are discrete and use an epsilon cutoff to counter rounding errors